### DISCUSSION TITLE
Uploaded Custom ESP32 Firmware But Does Not Work

### DISCUSSION THREAD

**Discussion by sir-user** [micropython/micropython#12717] [ESP32]
Hello to everyone,

I made a custom firmware for the ESP32 as I aimed to just add a few py modules to the ROM, but when I upload it for the first time. It worked fine, when I typed help('modules') I saw my modules in Thonny. Then I accidentally reuploaded it and didn't change anything but it didn't work.

This is the Thonny console output when I press the reset button on my custom ESP32 board. After this output, I can not reach my files or I can not upload code on Thonny. Device just waits.

Thank you in advance for your help,
Volkan

> ets Jul 29 2019 12:21:46
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4892
ho 0 tail 12 room 4
load:0x40078000,len:15092
load:0x40080400,len:4
load:0x40080404,len:3372
entry 0x400805c0
I (698) cpu_start: Multicore app
D (698) cpu_start: Pro cpu up
D (698) cpu_start: Starting app cpu, entry point is 0x400812e8
D (0) cpu_start: App cpu up
D (707) clk: RTC_SLOW_CLK calibration value: 2989069
I (719) cpu_start: Pro cpu start user code
I (719) cpu_start: cpu freq: 240000000 Hz
I (719) cpu_start: Application information:
I (724) cpu_start: Compile time:     Oct 17 2023 10:25:29
I (730) cpu_start: ELF file SHA256:  4033eab6d...
I (735) cpu_start: ESP-IDF:          v5.2-dev-3287-g1ec70cd568-dirty
I (743) cpu_start: Min chip rev:     v0.0
I (747) cpu_start: Max chip rev:     v3.99 
I (752) cpu_start: Chip rev:         v3.1
D (757) memory_layout: Checking 13 reserved memory ranges:
D (762) memory_layout: Reserved memory range 0x3ff81ff0 - 0x3ff82000
D (769) memory_layout: Reserved memory range 0x3ffae000 - 0x3ffae6e0
D (775) memory_layout: Reserved memory range 0x3ffae6e0 - 0x3ffaff10
D (782) memory_layout: Reserved memory range 0x3ffb0000 - 0x3ffb6388
D (788) memory_layout: Reserved memory range 0x3ffb8000 - 0x3ffb9a20
D (794) memory_layout: Reserved memory range 0x3ffbdb28 - 0x3ffbdb5c
D (801) memory_layout: Reserved memory range 0x3ffbdb60 - 0x3ffcb050
D (807) memory_layout: Reserved memory range 0x3ffe0000 - 0x3ffe0440
D (814) memory_layout: Reserved memory range 0x3ffe3f20 - 0x3ffe4350
D (820) memory_layout: Reserved memory range 0x40070000 - 0x40078000
D (827) memory_layout: Reserved memory range 0x40078000 - 0x40080000
D (833) memory_layout: Reserved memory range 0x40080000 - 0x4009ffec
D (839) memory_layout: Reserved memory range 0x50001fe8 - 0x50002000
D (846) memory_layout: Building list of available memory regions:
D (852) memory_layout: Available memory region 0x3ffaff10 - 0x3ffb0000
D (859) memory_layout: Available memory region 0x3ffb6388 - 0x3ffb8000
D (865) memory_layout: Available memory region 0x3ffb9a20 - 0x3ffbdb28
D (872) memory_layout: Available memory region 0x3ffcb050 - 0x3ffcc000
D (878) memory_layout: Available memory region 0x3ffcc000 - 0x3ffce000
D (885) memory_layout: Available memory region 0x3ffce000 - 0x3ffd0000
D (892) memory_layout: Available memory region 0x3ffd0000 - 0x3ffd2000
D (898) memory_layout: Available memory region 0x3ffd2000 - 0x3ffd4000
D (905) memory_layout: Available memory region 0x3ffd4000 - 0x3ffd6000
D (911) memory_layout: Available memory region 0x3ffd6000 - 0x3ffd8000
D (918) memory_layout: Available memory region 0x3ffd8000 - 0x3ffda000
D (925) memory_layout: Available memory region 0x3ffda000 - 0x3ffdc000
D (931) memory_layout: Available memory region 0x3ffdc000 - 0x3ffde000
D (938) memory_layout: Available memory region 0x3ffde000 - 0x3ffe0000
D (944) memory_layout: Available memory region 0x3ffe0440 - 0x3ffe3f20
D (951) memory_layout: Available memory region 0x3ffe4350 - 0x3ffe8000
D (958) memory_layout: Available memory region 0x3ffe8000 - 0x3fff0000
D (964) memory_layout: Available memory region 0x3fff0000 - 0x3fff8000
D (971) memory_layout: Available memory region 0x3fff8000 - 0x3fffc000
D (977) memory_layout: Available memory region 0x3fffc000 - 0x40000000
D (984) memory_layout: Available memory region 0x4009ffec - 0x400a0000
I (991) heap_init: Initializing. RAM available for dynamic allocation:
D (998) heap_init: New heap initialised at 0x3ffaff10
I (1003) heap_init: At 3FFAFF10 len 000000F0 (0 KiB): DRAM
D (1009) heap_init: New heap initialised at 0x3ffb6388
I (1014) heap_init: At 3FFB6388 len 00001C78 (7 KiB): DRAM
D (1020) heap_init: New heap initialised at 0x3ffb9a20
I (1026) heap_init: At 3FFB9A20 len 00004108 (16 KiB): DRAM
D (1032) heap_init: New heap initialised at 0x3ffcb050
I (1037) heap_init: At 3FFCB050 len 00014FB0 (83 KiB): DRAM
I (1043) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (1050) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (1056) heap_init: At 4009FFEC len 00000014 (0 KiB): IRAM
D (1063) intr_alloc: Connected src 46 to int 2 (cpu 0)
D (1068) spi_flash: trying chip: issi
D (1071) spi_flash: trying chip: gd
D (1075) spi_flash: trying chip: mxic
D (1079) spi_flash: trying chip: winbond
D (1083) spi_flash: trying chip: generic
I (1087) spi_flash: detected chip: generic
I (1091) spi_flash: flash io: dio
D (1095) cpu_start: calling init function: 0x401d9850
D (1101) cpu_start: calling init function: 0x401d9330
D (1106) cpu_start: calling init function: 0x4017fa7c
D (1111) cpu_start: calling init function: 0x4015a198
D (1116) cpu_start: calling init function: 0x40159440
D (1121) cpu_start: calling init function: 0x40151470
W (1126) rmt(legacy): legacy driver is deprecated, please migrate to `driver/rmt_tx.h` and/or `driver/rmt_rx.h`
D (1137) cpu_start: calling init function: 0x4014ece0
W (1142) i2s(legacy): legacy i2s driver is deprecated, please migrate to use driver/i2s_std.h, driver/i2s_pdm.h or driver/i2s_tdm.h
D (1154) cpu_start: calling init function: 0x4014a5fc
W (1160) i2c: This driver is an old driver, please migrate your application code to adapt `driver/i2c_master.h`
D (1170) cpu_start: calling init function: 0x40148ffc
W (1176) DAC: legacy driver is deprecated, please migrate to `driver/dac_oneshot.h`, `driver/dac_cosine.h` or `driver/dac_continuous.h` instead
D (1189) cpu_start: calling init function: 0x40148e3c
W (1194) ADC: legacy driver is deprecated, please migrate to `esp_adc/adc_oneshot.h`
D (1203) cpu_start: calling init function: 0x40097e84
D (1208) cpu_start: calling init function: 0x400d1f7c
D (1213) cpu_start: calling init function: 0x400d4d24 on core: 0
D (1219) intr_alloc: Connected src 17 to int 3 (cpu 0)
D (1224) cpu_start: calling init function: 0x400d2f28 on core: 0
I (1242) coexist: coex firmware version: 8770c12
D (1242) intr_alloc: Connected src 24 to int 9 (cpu 0)
D (1243) app_start: Starting scheduler on CPU0
 (1248) intr_alloc: Connected src 25 to int 2 (cpu 1)
D (1248) app_start: Starting scheduler on CPU1
2mI (1248) main_task: Started on CPU0
D (1268) heap_init: New heap initialised at 0x3ffe0440
D (1268) heap_init: New heap initialised at 0x3ffe4350
D (1268) intr_alloc: Connected src 16 to int 12 (cpu 0)
I (1268) main_task: Calling app_main

**Comment by shariltumin** [comment]
Is it possible that you have reflashed something other than the MicroPython firmware?

```
D (1268) intr_alloc: Connected src 16 to int 12 (cpu 0)
I (1268) main_task: Calling app_main
```
does not look like the board has loaded and executed MicroPython.

Please try erasing and reflashing the board using **esptool.py**.

Erase flash
```
esptool.py --port <com-port> erase_flash
```

Reflash firmware
```
esptool.py --port  <com-port> --chip esp32 --baud 460800 write_flash -z 0x1000 <firmware.bin>
```

Please replace <com-port> and <firmware.bin> with your values.

**Reply by shariltumin** [reply]
Please replace <com-port>  and <firmware.bin> with your values.

**Reply by sir-user** [reply]
I tried it but does not work, still same output. 
Then I deleted all files and reinstall Micropython&ESP-IDF but I got iram0_0_seg overflow 1228 bytes.  (I didn't add any files and even removed some modules by manifest., it is just micropython files)

I saw a solution for overflow error on another post and I use it but I got` rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)`, spıwp:0xee error. When I power on the device, I get in this error loop. İt continuesly SPI_FAST_FLASH_BOOT reset.

**Reply by shariltumin** [reply]
You can try downloading a firmware from micropython.org, flash the official firmware and see if it works. This is to test that nothing is wrong with your board.

**Reply by sir-user** [reply]
it works, I uploaded ESP32_GENERIC firmware to my custom ESP32 device from micropython.org.

Why did I get overflow error? I even removed some modules by manifest. 
When I type idf.py size:
```
Total sizes:
Used static DRAM:   54508 bytes (  70072 remain, 43.8% used)
      .data size:   22244 bytes
      .bss  size:   32264 bytes
Used static IRAM:  132298 bytes (  -1226 remain, 100.9% used) Overflow detected! You can run idf.py size-files for more information.
      .text size:  131271 bytes
   .vectors size:    1027 bytes
Used Flash size : 1547059 bytes
           .text: 1276039 bytes
         .rodata:  270764 bytes
Total image size: 1701601 bytes (.bin may be padded larger)
```

**Reply by shariltumin** [reply]
Now that you have MP up and running, you can try to import your modules from the REPL.

**Reply by sir-user** [reply]
@shariltumin I want to add some modules to ROM but I got overflow error even I removed some modules by manifest.py. I want to add my modules but I get bigger overflow error when I add some. 
Is there any way to minimazing size? When I [this document](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/performance/ram-usage.html#optimizing-iram-usage) implement it to my own code, I build it well but I get  `rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)` error and device gets in loop.

**Reply by karfas** [reply]
I think you should try first to build from a fresh, unmodified clone of micropython using the ESP-IDF (5.0.2 ?) also used to build the official firmware.
When this works,  you can gradually add/remove the modules you require.

At some time months ago I got a similar error after a git rebase to the current master.
As far as I remember, I had either missed the git submodule sync + git submodule update or I had to remove the build directory to get my build working.

**Reply by shariltumin** [reply]
I do not know what are necessary to include in your custom firmware.

I usually have to choose what to include in my custom firmware.

I exclude help, webrepl, espnow and if necessary I also exclude asyncio from fozen modules.

If that is not enough, I also exclude BLE.

This can be done by disabling flags in ports/esp32/mpconfigport.h for example
```
#define MICROPY_PY_BUILTINS_HELP            (0)
#define MICROPY_PY_BUILTINS_HELP_MODULES    (0)
#define MICROPY_PY_WEBREPL                  (0)
#define MICROPY_PY_SOCKET_EVENTS            (1)
#define MICROPY_ESPNOW                      (0)
```

It all depends on what you want to use your custom Micropython for.

I know next to nothing about tuning IRAM usage.

**Comment by sir-user** [comment]
Hi guys,

Thanks a lot for replies, 
I solved the problem, now I can add my files and remove. I guess my Micropython and ESP-IDF versions was not match. Then I deleted ESP-IDF and Micropython files and then reinstall them a few times until it matches.
